Video player refactoring to more idiomatic Kotlin

elizarov opened the pull request:
Got rid of most C-style nullable refs, C-style memory management & vars.
Significatnly reduced the usage of '!!' operator.
Introduced "Disposable" DSL to declare owned resources that shall be freed.
Disposable completely replaces C-style 'init' & 'deinit' methods.
Resource refs aren kept in non-nullable 'vals' instead of  nullable 'vars'.
Properly close av file on exit & support video/audio-only modes for testing.

Also, A/V sync code is improved. It maintains perfect FPS if possible with or without sound sync.
TODO: Shall use frame timestamps for smooth A/V sync. It is good enough for now, though. 
Video is now smooth, without periodic halts on resyncs with audio.

olonho commented on samples/videoplayer/src/main/kotlin/Queue.kt:
```diff
@@ -14,10 +14,10 @@
-    val array = Array<T>(maxSize, { _ -> none})
-    var head = 0
-    var tail = 0
+class Queue<T>(val maxSize: Int) {
```
Эти изменения вряд ли что-то улучшают.

elizarov replied:
This change is to make this `Queue` impl behave like any other idiomatic Kotlin collection (like ArrayList, HashSet, etc). Previous impl of `Queue` forced the user to use nullable types.

For example, `videoQueue` was declared as `Queue<VideoFrame?>` while, in fact, `null` frame were never pushed to queue, so the type was misleading to anyone reading the code. I, frankly, spend some time figuring out why the element type is nullable, since I could not find any code that adds a null frame into this queue.

With this change it is used with non-null types. `videoQueue` is now declared as `Queue<VideoFrame>`. It behaves just like all other Kotlin collections now.

olonho commented on samples/videoplayer/src/main/kotlin/Disposable.kt:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    // These functions should be used in constructor only.
+    // They will properly dispose of partially initialized object on failure
```
Точка в конце предложения.  Здесь как-то очень абстрактные понятия и SDL смешаны не совсем по делу,

elizarov replied:
Fixed. Improved documentation on `Disposable`

olonho commented on samples/videoplayer/src/main/kotlin/Disposable.kt:
```diff
@@ -0,0 +1,47 @@
+    // These functions should be used in constructor only.
+    // They will properly dispose of partially initialized object on failure
+
+    inline fun <T> tryConstruct(init: () -> T): T =
```
:T наверное лишнее.

elizarov replied:
This is a common pattern to pass-through the result in kotlin scoping functions. In particular, it is used in the following places: 
https://github.com/elizarov/kotlin-native/blob/920e9629a1d6e14deb6c4238ba959fb7315fe084/samples/videoplayer/src/main/kotlin/Disposable.kt#L31
https://github.com/elizarov/kotlin-native/blob/re/video-player-refactoring/samples/videoplayer/src/main/kotlin/SDLVideo.kt#L28

olonho commented on samples/videoplayer/src/main/kotlin/SDLAudio.kt:
```diff
@@ -16,85 +16,72 @@
 import platform.posix.memcpy
 
-class SDLAudio(val player: VideoPlayer) : SDLBase() {
+class SDLAudio(val player: VideoPlayer) : DisposableContainer() {
```
Может проще было просто Arena передавать от плеера?

elizarov replied:
Conceptually, `Disposable` class has to be fully responsible for its resource management. The lifecycle is simple: You create, then you invoke it, then you `dispose` on it.

The code was refactored in a such a way, as to make sure that each resource-containing class if fully responsible for the resources it allocated, so there is no need to keep this information in mind when programming. There is never a confusion what class "owns" resources. They are always owned by the class that has the corresponding field.  This simple resource-management convention makes it very easy to ensure that no resource is lost.

elizarov replied:
I've documented the convention in `Disposable` interface.

olonho commented on samples/videoplayer/src/main/kotlin/SDLErrors.kt:
```diff
@@ -17,9 +17,15 @@
-    protected fun get_SDL_Error() = SDL_GetError()!!.toKString()
+private fun logSDLError(name: String): String {
+    val msg = "SDL_$name Error: ${SDL_GetError()!!.toKString()}"
+    println(msg)
```
println(msg) лишнее

elizarov replied:
Fixed. Simplified code.

olonho commented on samples/videoplayer/src/main/kotlin/Vec2D.kt:
```diff
@@ -0,0 +1,5 @@
+
+data class Vec2D(val x: Int, val y: Int) {
```
Vec2D -> Vector2D

elizarov replied:
Renamed to `Dimensions(w,h)` based on the below comment.

olonho commented on samples/videoplayer/src/main/kotlin/VideoPlayer.kt:
```diff
@@ -22,7 +22,13 @@ import platform.posix.*
+    inline fun transition(from: State, to: State, block: () -> Unit): State =
+        if (this == from) {
+            block()
+            to
```
|`line 90`|
|---|
Это все же Dimensions, или Rectangle а не вектор.

elizarov replied:
Renamed

olonho commented on samples/videoplayer/src/main/kotlin/SDLAudio.kt:
```diff
@@ -16,85 +16,72 @@
 import platform.posix.memcpy
 
-class SDLAudio(val player: VideoPlayer) : SDLBase() {
+class SDLAudio(val player: VideoPlayer) : DisposableContainer() {
```
Я примерно и пытался донести мысль, что можно атрибутировать ресурсы тому, кто контролирует лайфтайм этих объектов, путем передачи  некоторого lifetime manager'а, который суть Arena.

olonho commented on samples/videoplayer/src/main/kotlin/Queue.kt:
```diff
@@ -14,10 +14,10 @@
-    val array = Array<T>(maxSize, { _ -> none})
-    var head = 0
-    var tail = 0
+class Queue<T>(val maxSize: Int) {
```
Насчет нулабилити наружу это да, но я думаю того же эффекта можно добиться без некрасивых кастов на каждый доступ к очереди. То что я написал можно нормально будет специализировать, такой код - нет. Короче, лучше оставить как есть, возможно чуть почище обработав нулабилити.

olonho commented:
Tried this PR, and audio is broken in samples I use, such as http://www.sample-videos.com/video/mp4/720/big_buck_bunny_720p_10mb.mp4

elizarov commented:
I've fixed the audio problem (see the latest commit). Tested -- it works. However, I've also rebased it on top of the master and it does not build anymore. `gradlew :videoplayer:build` on master results in:

```
llvm-lto: error: Linking globals named 'Konan_main': symbol multiply defined!
exception: org.jetbrains.kotlin.backend.konan.KonanExternalToolFailure: The /Users/jetbrains/.konan/dependencies/clang-llvm-5.0.0-darwin-macos/bin/llvm-lto command returned non-zero exit code: 1.
        at org.jetbrains.kotlin.backend.konan.LinkStage.runTool(LinkStage.kt:475)
        at org.jetbrains.kotlin.backend.konan.LinkStage.llvmLto(LinkStage.kt:332)
        at org.jetbrains.kotlin.backend.konan.LinkStage.access$llvmLto(LinkStage.kt:287)
        at org.jetbrains.kotlin.backend.konan.LinkStage$linkStage$2.invoke(LinkStage.kt:498)
        at org.jetbrains.kotlin.backend.konan.LinkStage$linkStage$2.invoke(LinkStage.kt:287)
        at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:137)
        at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:116)
        at org.jetbrains.kotlin.backend.konan.util.UtilKt.profileIf(util.kt:33)
        at org.jetbrains.kotlin.backend.konan.PhaseManager.phase$backend_native_compiler(KonanPhases.kt:136)
        at org.jetbrains.kotlin.backend.konan.LinkStage.linkStage(LinkStage.kt:493)
        at org.jetbrains.kotlin.backend.konan.KonanDriverKt$runTopLevelPhases$5.invoke(KonanDriver.kt:106)
        at org.jetbrains.kotlin.backend.konan.KonanDriverKt$runTopLevelPhases$5.invoke(KonanDriver.kt)
        at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:137)
        at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:116)
        at org.jetbrains.kotlin.backend.konan.util.UtilKt.profileIf(util.kt:33)
        at org.jetbrains.kotlin.backend.konan.PhaseManager.phase$backend_native_compiler(KonanPhases.kt:136)
        at org.jetbrains.kotlin.backend.konan.KonanDriverKt.runTopLevelPhases(KonanDriver.kt:105)
        at org.jetbrains.kotlin.cli.bc.K2Native.doExecute(K2Native.kt:55)
        at org.jetbrains.kotlin.cli.bc.K2Native.doExecute(K2Native.kt:41)
        at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.java:107)
        at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.java:51)
        at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:92)
        at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:70)
        at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:36)
        at org.jetbrains.kotlin.cli.common.CLITool$Companion.doMainNoExit(CLITool.kt:157)
        at org.jetbrains.kotlin.cli.common.CLITool$Companion.doMain(CLITool.kt:148)
        at org.jetbrains.kotlin.cli.bc.K2Native$Companion$main$1.invoke(K2Native.kt:168)
        at org.jetbrains.kotlin.cli.bc.K2Native$Companion$main$1.invoke(K2Native.kt:159)
        at org.jetbrains.kotlin.backend.konan.util.UtilKt.profileIf(util.kt:33)
        at org.jetbrains.kotlin.backend.konan.util.UtilKt.profile(util.kt:28)
        at org.jetbrains.kotlin.cli.bc.K2Native$Companion.main(K2Native.kt:161)
        at org.jetbrains.kotlin.cli.bc.K2NativeKt.main(K2Native.kt:173)
        at org.jetbrains.kotlin.cli.utilities.MainKt.main(main.kt:115)
```

elizarov commented:
Additional comment: None of the example now build on my Mac Book in master branch, so I assume that videoplayer itself is Ok.

olonho commented:
Попробуй `./gradlew clean`

olonho commented on samples/videoplayer/src/main/kotlin/DecoderWorker.kt:
```diff
@@ -15,387 +15,394 @@
-                resampledAudioFrame!!.pointed.nb_samples,
-                resampledAudioFrame!!.pointed.format,
-                1)
-        val ts = av_frame_get_best_effort_timestamp(audioFrame) * av_q2d(
```
Это не лучший фикс, как мне кажется. Я специально транскодировал все в 44100/S16/2 чтобы SDL всегда это мог сыграть без дополнительного второго транскодирования в зависимости от звукового бекенда. Лучше переделать на такой подход.

elizarov commented:
Rebased once more to master, rebuilt clean, all works. Sound if working property now. 

Sound is always resampled to 2 stereo channels at 44100 Hz as it was before, but this  magic constant for 44100 is not mentioned in multiple files anymore. This internal playback format is defined in one place now: https://github.com/elizarov/kotlin-native/blob/a7f9f14ea3db04ac5c935a083d2aa028704a5c4a/samples/videoplayer/src/main/kotlin/VideoPlayer.kt#L90

olonho commented:
Merged as https://github.com/JetBrains/kotlin-native/pull/1190
